2745: Upgrade react-native

steffenkleinle opened the pull request:
### Short description

<!-- Describe this PR in one or two sentences. -->
Upgrade react-native and dependencies blocked by this upgrade.

### Proposed changes

<!-- Describe this PR in more detail. -->

- Upgrade react-native according to https://react-native-community.github.io/upgrade-helper/?from=0.71.12&to=0.73.3
- Upgrade react-native-firebase to its latest version
- Migrate our native modules to kotlin
- Tidy up android build files
- Use `^` for all dependencies not blocked
- Tidy up the package.jsons a bit (aka remove unused/unnecessary dependencies)
- Tidy up android resources and remove custom splash screen

### Side effects

<!-- List all related components that have not been explicitly changed but may be affected by this PR -->
Hopefully none. 

### Known issues

- [x] ~~Image size is off~~ styled-component style output format has changed (https://github.com/digitalfabrik/integreat-app/pull/2757/commits/fb1c1bd346679a67e54a68b64b62c56926a57760)
- [x] ~~Warning: No task registered for key ReactNativeFirebaseMessagingHeadlessTask~~: Could not reproduce it anymore
- [x] ~~Could not invoke RCTMGLLocationModule.getLastKnownLocation (maps)~~ Downgrade @react-native-community/geolocation (https://github.com/digitalfabrik/integreat-app/pull/2757/commits/4d3970f23b9d19cb8e6b80e30590f5360bc13cf8)

### Resolved issues

<!-- List all issues which should be closed when this PR is merged. -->

Fixes: #2745.

---

<!--
DOR:
- [Release notes](https://github.com/digitalfabrik/integreat-app/blob/main/docs/contributing.md#release-notes) have been added for user visible changes
- Linting: `yarn lint`
- Typescript: `yarn ts:check`
- Prettier: `yarn prettier`
- Unit tests: `yarn test`
- -->

steffenkleinle commented on native/package.json:
```diff
@@ -50,10 +50,10 @@
     "@react-native-async-storage/async-storage": "^1.23.1",
     "@react-native-clipboard/clipboard": "^1.14.0",
-    "@react-native-community/geolocation": "^3.2.1",
+    "@react-native-community/geolocation": "3.0.6",
```
I had to downgrade because of this (or a similar) error:
https://github.com/rnmapbox/maps/issues/3046

steffenkleinle commented:
Are push notifications working for you?

LeandraH reviewed (approved):
Tested in real and emulated iOS and emulated Android, looks good! The Integreat icon in the header is a little off but that's the only thing I found. And the notification worked too.

steffenkleinle commented on native/android/.idea/compiler.xml:
Some files in the .idea folder are already gitignored. Some also relate to project config/how to open a project. In this case I think it makes sense to keep those in git.

steffenkleinle commented on native/android/settings.gradle:
```diff
@@ -3,4 +3,4 @@ rootProject.name = 'integreat-app-android'
 include ':app'
-includeBuild('../node_modules/react-native-gradle-plugin')
\ No newline at end of file
+includeBuild('../node_modules/@react-native/gradle-plugin')
```
I wouldn't. This is all just taken from the official react-native upgrade helper. If this changes in the future, we will notice.

steffenkleinle commented on native/ios/Gemfile:
```diff
@@ -5,4 +5,7 @@ gem "fastlane", ">= 2.216.0"
 eval_gemfile(plugins_path) if File.exist?(plugins_path)
 
-gem "cocoapods", "~>1.11", "= 1.11.2"
+# Cocoapods 1.15 introduced a bug which break the build. We will remove the upper
```
Copied from react-native, so I assume there should be a ticket. This will probably be changed in an upcoming react-native version.

steffenkleinle commented on native/tsconfig.json:
```diff
@@ -16,7 +16,7 @@
     },
     "skipLibCheck": true,
     "typeRoots": ["./src/@types", "./node_modules/@types", "../node_modules/@types"],
-    "types": ["jest", "node", "react-native"]
```
react-native got rewritten in typescript/the default is now typescript and therefore it is not needed anymore. An error is thrown is this is not removed.

steffenkleinle commented on native/src/components/base/Icon.tsx:
```diff
@@ -12,7 +12,7 @@ type IconProps = {
   directionDependent?: boolean
   reverse?: boolean
-  style?: Record<string, number>[]
+  style?: { width?: number; height?: number; color?: ColorValue }
```
Either styled-components or react-native changed something in the way styles are passed through. Before an array was passed, now its just an object. If you revert those changes, the styles are not applied correctly anymore.

steffenkleinle commented:
@LeandraH thank you! :green_heart: 

> Tested in real and emulated iOS and emulated Android, looks good! The Integreat icon in the header is a little off but that's the only thing I found. And the notification worked too.

Did you test push notifications in all states (foreground, background, quit)? If not, could you please do that? I am a little afraid of the warning mentioned in the PR description:
> Warning: No task registered for key ReactNativeFirebaseMessagingHeadlessTask

LeandraH commented:
I tested foreground and background yesterday and it worked perfectly, I tested quit state just now and it opened the app but on the dashboard. That might be because the dev app on my phone wasn't connected with Xcode anymore, though, I'm not quite sure how to even test that :(

ztefanie reviewed (commented):
:wrench: Please adjust the node version in the README.md (i needed to at least use 20 to succeed with installing).

I made an Android build with fastlane and it did not succeed:
`/home/steffi/Dokumente/integreat-app/native/android/app/src/main/res/drawable-land-hdpi/launch_screen.png: Error: The drawable "launch_screen" in drawable-land-hdpi has no declaration in the base drawable folder or in a drawable-densitydpi folder; this can lead to crashes when the drawable is queried in a configuration that does not match this qualifier [MissingDefaultResource]`

Can you build with fastlane without errors?

steffenkleinle commented on native/android/build.gradle:
```diff
@@ -1,58 +1,20 @@
     }
     dependencies {
-        classpath("com.android.tools.build:gradle:7.3.1")
+        classpath("com.android.tools.build:gradle")
```
I just followed the official upgrade instructions from https://react-native-community.github.io/upgrade-helper/?from=0.71.12&to=0.73.7. So I don't see a reason not to remove it.

steffenkleinle commented:
> ❌ I made an Android build with fastlane and it did not succeed: `/home/steffi/Dokumente/integreat-app/native/android/app/src/main/res/drawable-land-hdpi/launch_screen.png: Error: The drawable "launch_screen" in drawable-land-hdpi has no declaration in the base drawable folder or in a drawable-densitydpi folder; this can lead to crashes when the drawable is queried in a configuration that does not match this qualifier [MissingDefaultResource]`
> 
> Can you build with fastlane without errors?

Will try.

f1sh1918 reviewed (commented):
Ios works fine but i get an error on android when i try to build the app on emulator

```
error Failed to install the app. Looks like your Android environment is not properly set. Please go to https://reactnative.dev/docs/0.73/environment-setup?os=macos&platform=android&guide=native#jdk-studio and follow the React Native CLI QuickStart guide to install the compatible version of JDK.
info Run CLI with --verbose flag for more details.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
node:child_process:965
    throw err;
    ^

Error: Command failed: yarn cross-env BUILD_CONFIG_APP_NAME="IntegreatTestCms" BUILD_CONFIG_APP_ICON="app_icon_integreat" BUILD_CONFIG_NOTIFICATION_ICON="notification_icon_integreat" BUILD_CONFIG_LIGHT_THEME_COLORS_THEME_COLOR="#fbda16" BUILD_CONFIG_LIGHT_THEME_COLORS_THEME_COLOR_LIGHT="rgba(251,_218,_22,_0.5)" BUILD_CONFIG_LIGHT_THEME_COLORS_BACKGROUND_ACCENT_COLOR="#fafafa" BUILD_CONFIG_LIGHT_THEME_COLORS_TEXT_COLOR="#000000" BUILD_CONFIG_LIGHT_THEME_COLORS_TEXT_SECONDARY_COLOR="#585858" BUILD_CONFIG_LIGHT_THEME_COLORS_TEXT_DECORATION_COLOR="#c7c7c7" BUILD_CONFIG_LIGHT_THEME_COLORS_TEXT_DISABLED_COLOR="#d0d0d0" BUILD_CONFIG_LIGHT_THEME_COLORS_FOOTER_LINE_COLOR="#b1b1b1" BUILD_CONFIG_LIGHT_THEME_COLORS_BACKGROUND_COLOR="#ffffff" BUILD_CONFIG_LIGHT_THEME_COLORS_TUNEWS_THEME_COLOR="#007aa8" BUILD_CONFIG_LIGHT_THEME_COLORS_TUNEWS_THEME_COLOR_LIGHT="rgba(0,_122,_168,_0.4)" BUILD_CONFIG_LIGHT_THEME_COLORS_BORDER_COLOR="#F1F1F1" BUILD_CONFIG_LIGHT_THEME_COLORS_POSITIVE_HIGHLIGHT="#188038" BUILD_CONFIG_LIGHT_THEME_COLORS_NEGATIVE_HIGHLIGHT="#8b0000" BUILD_CONFIG_LIGHT_THEME_COLORS_INVALID_INPUT="#B3261E" BUILD_CONFIG_LIGHT_THEME_FONTS_NATIVE_DECORATIVE_FONT_BOLD="Raleway-Bold" BUILD_CONFIG_LIGHT_THEME_FONTS_NATIVE_DECORATIVE_FONT_REGULAR="Raleway-Regular" BUILD_CONFIG_LIGHT_THEME_FONTS_NATIVE_CONTENT_FONT_REGULAR="NotoSans" BUILD_CONFIG_LIGHT_THEME_FONTS_NATIVE_CONTENT_FONT_BOLD="NotoSans-Bold" BUILD_CONFIG_LIGHT_THEME_FONTS_NATIVE_WEBVIEW_FONT="Noto_Sans,_Noto_Sans_Arabic,_Noto_Sans_SC,_Noto_Sans_Georgian,_sans-serif" BUILD_CONFIG_LIGHT_THEME_FONTS_WEB_DECORATIVE_FONT="Raleway,_Noto_Sans_Arabic,_Noto_Sans_SC,_Noto_Sans_Georgian,_sans-serif" BUILD_CONFIG_LIGHT_THEME_FONTS_WEB_CONTENT_FONT="Noto_Sans,_Noto_Sans_Arabic,_Noto_Sans_SC,_Noto_Sans_Georgian,_sans-serif" BUILD_CONFIG_LIGHT_THEME_FONTS_FONT_SIZE_ADJUST="0.55" BUILD_CONFIG_LIGHT_THEME_FONTS_DECORATIVE_FONT_SIZE="0.95rem" BUILD_CONFIG_LIGHT_THEME_FONTS_DECORATIVE_FONT_SIZE_SMALL="0.7rem" BUILD_CONFIG_LIGHT_THEME_FONTS_DECORATIVE_LINE_HEIGHT="1.3" BUILD_CONFIG_LIGHT_THEME_FONTS_CONTENT_FONT_SIZE="0.95rem" BUILD_CONFIG_LIGHT_THEME_FONTS_CONTENT_LINE_HEIGHT="1.4" BUILD_CONFIG_LIGHT_THEME_FONTS_STANDARD_PARAGRAPH_MARGIN="0.75rem" BUILD_CONFIG_LIGHT_THEME_FONTS_SUB_TITLE_FONT_SIZE="1.125rem" BUILD_CONFIG_LIGHT_THEME_FONTS_HINT_FONT_SIZE="0.85rem" BUILD_CONFIG_LIGHT_THEME_FONTS_ADAPTIVE_FONT_SIZE_SMALL_MIN="0.55rem" BUILD_CONFIG_LIGHT_THEME_FONTS_ADAPTIVE_FONT_SIZE_SMALL_VALUE="1.6vh" BUILD_CONFIG_LIGHT_THEME_FONTS_ADAPTIVE_FONT_SIZE_SMALL_MAX="0.85rem" BUILD_CONFIG_ASSETS="integreat" BUILD_CONFIG_CMS_URL="https://cms-test.integreat-app.de" BUILD_CONFIG_SWITCH_CMS_URL="https://cms.integreat-app.de" BUILD_CONFIG_HOST_NAME="integreat.app" BUILD_CONFIG_ALLOWED_HOST_NAMES_0="cms.integreat-app.de" BUILD_CONFIG_ALLOWED_HOST_NAMES_1="cms-test.integreat-app.de" BUILD_CONFIG_ALLOWED_HOST_NAMES_2="admin.integreat-app.de" BUILD_CONFIG_ALLOWED_LOOKALIKES_0="https://integreat.app" BUILD_CONFIG_ALLOWED_LOOKALIKES_1="https://integreat-app.de" BUILD_CONFIG_SUPPORTED_IFRAME_SOURCES_0="vimeo.com" BUILD_CONFIG_INTERNAL_LINKS_HIJACK_PATTERN="https?:\/\/(cms(-test)?\.integreat-app\.de|web\.integreat-app\.de|integreat\.app)(?!\/(media|[^/]*\/(wp-content|wp-admin|wp-json))\/.*).*" BUILD_CONFIG_FEATURE_FLAGS_FLOSS="false" BUILD_CONFIG_FEATURE_FLAGS_POIS="true" BUILD_CONFIG_FEATURE_FLAGS_NEWS_STREAM="true" BUILD_CONFIG_FEATURE_FLAGS_PUSH_NOTIFICATIONS="true" BUILD_CONFIG_FEATURE_FLAGS_INTRO_SLIDES="true" BUILD_CONFIG_FEATURE_FLAGS_JPAL_TRACKING="true" BUILD_CONFIG_FEATURE_FLAGS_SENTRY="false" BUILD_CONFIG_FEATURE_FLAGS_DEVELOPER_FRIENDLY="true" BUILD_CONFIG_FEATURE_FLAGS_FIXED_CITY="null" BUILD_CONFIG_FEATURE_FLAGS_CITY_NOT_COOPERATING_TEMPLATE="Sehr_geehrte_Damen_und_Herren,\n\nIch_wohne_seit_einiger_Zeit_in_[Ort]_und_habe_neulich_versucht,_meinen_Ort_in_der_Integreat-App_zu_finden._Die_App_stellt_zahlreiche_wichtige_Informationen_über_den_Alltag_und_die_Region_in_verschiedenen_Sprachen_zur_Verfügung._Das_erleichtert_vielen_Neuzugezogenen_die_Eingewöhnung,_was_ich_sehr_hilfreich_finde.\n\nIch_würde_mich_daher_sehr_freuen,_wenn_auch_Sie_zukünftig_bei_Integreat_mitmachen.\n\n\n\nAuf_der_Webseite_erfahren_Sie_mehr:\n\nhttps://integreat-app.de\n\n\n\nHerzliche_Grüße" BUILD_CONFIG_FEATURE_FLAGS_CITY_NOT_COOPERATING="true" BUILD_CONFIG_ABOUT_URLS_DEFAULT="https://integreat-app.de/about/" BUILD_CONFIG_ABOUT_URLS_EN="https://integreat-app.de/en/about/" BUILD_CONFIG_PRIVACY_URLS_DEFAULT="https://integreat-app.de/datenschutz/" BUILD_CONFIG_PRIVACY_URLS_EN="https://integreat-app.de/en/privacy/" BUILD_CONFIG_SPLASH_SCREEN="true" BUILD_CONFIG_APPLICATION_ID="app.integreat.test" BUILD_CONFIG_GOOGLE_SERVICES_GOOGLE_APP_ID="1:164298278764:android:a0847ab9a13225b66fd549" BUILD_CONFIG_GOOGLE_SERVICES_GCM_DEFAULT_SENDER_ID="164298278764" BUILD_CONFIG_GOOGLE_SERVICES_DEFAULT_WEB_CLIENT_ID="164298278764-hhup7orh5rvg0oumhbabqvp5giudaste.apps.googleusercontent.com" BUILD_CONFIG_GOOGLE_SERVICES_GA_TRACKING_ID="null" BUILD_CONFIG_GOOGLE_SERVICES_FIREBASE_DATABASE_URL="https://integreat-2020.firebaseio.com" BUILD_CONFIG_GOOGLE_SERVICES_GOOGLE_API_KEY="AIzaSyBvc08_Cqww8z2Dh-EXtwHW98HXBk8T7V4" BUILD_CONFIG_GOOGLE_SERVICES_GOOGLE_CRASH_REPORTING_API_KEY="AIzaSyBvc08_Cqww8z2Dh-EXtwHW98HXBk8T7V4" BUILD_CONFIG_GOOGLE_SERVICES_PROJECT_ID="integreat-2020" BUILD_CONFIG_NAME="integreat-test-cms"  yarn react-native run-android --no-packager --appId app.integreat.test 
    at checkExecSyncError (node:child_process:890:11)
    at execSync (node:child_process:962:15)
    at Command.<anonymous> (/Users/afischer/projects/integreat-app/native/run:36:5)
    at Command.listener [as _actionHandler] (/Users/afischer/projects/integreat-app/native/node_modules/commander/lib/command.js:523:17)
    at /Users/afischer/projects/integreat-app/native/node_modules/commander/lib/command.js:1375:65
    at Command._chainOrCall (/Users/afischer/projects/integreat-app/native/node_modules/commander/lib/command.js:1272:12)
    at Command._parseCommand (/Users/afischer/projects/integreat-app/native/node_modules/commander/lib/command.js:1375:27)
    at /Users/afischer/projects/integreat-app/native/node_modules/commander/lib/command.js:1161:27
    at Command._chainOrCall (/Users/afischer/projects/integreat-app/native/node_modules/commander/lib/command.js:1272:12) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 38828,
  stdout: null,
  stderr: null
}

Node.js v20.11.1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

steffenkleinle commented:
> 🔧 Please adjust the node version in the README.md (i needed to at least use 20 to succeed with installing).

Done, good catch.

> ❌ I made an Android build with fastlane and it did not succeed: `/home/steffi/Dokumente/integreat-app/native/android/app/src/main/res/drawable-land-hdpi/launch_screen.png: Error: The drawable "launch_screen" in drawable-land-hdpi has no declaration in the base drawable folder or in a drawable-densitydpi folder; this can lead to crashes when the drawable is queried in a configuration that does not match this qualifier [MissingDefaultResource]`
> 
> Can you build with fastlane without errors?

Should be fixed. I updated all used icons and tidied up. I also removed the old custom launch screen as the default android one looks better (and the old one was probably also the reason for the error).
